### ISSUE TITLE
Feature Request - Decals that work with animated meshes

### ISSUE THREAD

**Issue by daarong** [mrdoob/three.js#7926]
Currently decals (DecalGeometry.js) do not work with animated meshes.

I am peering into the code to see how difficult it would be to implement this.  It appears maybe animations are done at the shader level, so I'm getting increasingly discouraged the more I'm trying to analyze this.

I'm sure many would appreciate this enhancement!

Thanks

**Comment by linkedz** [comment]
+1 (for SkinnedMesh)

**Comment by makc** [comment]
@daarong it should be quite simple, conceptually. you just need to copy/interpolate skin indices/weights along with other vertex data.

**Comment by daarong** [comment]
@makc I thought so as well.  I took a stab at it but the result was wildly swinging vertices.  Upon further thought I think I forgot to recreate the skin indices (I simply copied).  DecalGeometry produces a new mesh from existing vertices.  Those new vertices have different indices than the original mesh, so things do have to be re-calculated a bit.  Other than that hopefully it's just a matter of copying things.

**Comment by daarong** [comment]
@makc Yes, recreating the skinIndices rather than copying the original ones fixes it.

But now I remember why I was looking so deeply into the animation implementation: Once the mesh is deformed according to the animation, the vertices need to continue to be modified according to the animation throughout the process of making the decal (eg while identifying where the hit occurs and while creating the decal vertices).

I'm getting too far into the weeds... since this is not currently implemented I'm going to attempt a poor workaround that will allow me to do what I want without modifying / updating ThreeJS.

**Comment by makc** [comment]
I don't get it

> Once the mesh is deformed according to the animation, the vertices need to continue to be modified according to the animation throughout the process of making the decal (eg while identifying where the hit occurs and while creating the decal vertices).

are you saying that skin with decal is transformed differently than original skin? then you got your weights wrong.

**Comment by daarong** [comment]
No, that's not what I saw.  For example in the "skinning blending" demo you see the marine with his arms straight out.  Once any animation starts the arms are never straight out again.  However DecalGeometry looks at the 'raw' vertices with only basic mesh transformations applied (rot/pos/scale) but not with animation applied.

So if the marine is walking with his arms at his side, clicking an arm registers nothing because DecalGeometry expects the arms to still be straight outward in their un-animated locations (and indeed clicking in that empty space outside his shoulders does cause a decal to be generated).  I hope that clarifies!

**Comment by makc** [comment]
yes, of course, but I thought since you are adding skin weights/indices you will be also using SkinnedMesh (not plain Mesh) for decal with the same bones/animation that original mesh is using. then decal would just follow the skin naturally.

**Comment by daarong** [comment]
Yes! DecalGeometry produces a mesh... I fed that mesh along with decal material into the SkinnedMesh constructor.  Inside DecalGeometry I copied the bones/animations associated with the original.

I may be making it more complicated than necessary, but my swing at it was not a success

**Comment by makc** [comment]
Do you have this somewhere in your fork?

**Comment by daarong** [comment]
Sorry nothing useful.  I left that stuff behind (read:deleted) really quickly as I decided it wasn't worth pursuing.

In case somebody else wants a starting point to tinker, I basically did this:

1) combine "animation / skinning / blending" example with pieces of "decal" example
2) in DecalGeometry -> computeDecal, I collected vertex indices in the first for loop while the vertices themselves were collected
3) in DecalGeometry -> computeDecal, I copied animations and bones from original mesh
4) in DecalGeometry -> computeDecal, I created new skinIndices using the iterator value (k)
5) in DecalGeometry -> computeDecal, I copied the skinWeights from the original mesh geometry using the newly collected vertex indices
6) in the main HTML -> onMouseDown, I set "skinning" of decal material to true
7) in the main HTML -> onMouseDown, I altered the new geometry according to the original mesh so that scale, rotation, translation were 'reset' ... because of #8
8) in the main HTML -> onMouseDown, I added the new geometry as a child to the original mesh

That's essentially what I did, but as I said, it wasn't enough...

**Comment by makc** [comment]
>  I set "skinning" of decal material to true

it was not enough because the object has to actually be skinned mesh, not just mesh. and, unless you copy the bones by reference, you need to animate it too. probably.

also, when decalgeometry calls [this](https://github.com/mrdoob/three.js/blob/master/examples/js/geometries/DecalGeometry.js#L37) you need to interpolate skin weights correspondin to same skin indices.

**Comment by daarong** [comment]
DecalGeometry produces only a geometry... I did use that geometry in a SkinnedMesh constructor.  Also I did copy them be reference, and they did "animate", albeit it wasn't correct.

Yes, my problem may have been there... I never got to the point where I understood the clip or clipFace functions...

**Comment by Mugen87** [comment]
I vote to close this feature request and keep `DecalGeometry` simple. It seems the feature request was filed without understanding the implications that are introduced by the complex implementation that is necessary to support decals on animated surfaces.

**Comment by makc** [comment]
@Mugen87 well as far as noone is doing it, you can close it, of course, and it has been 2 years since the guy needed it. Some day someone could add AnimatedDecalGeometry with all that jazz.

**Comment by makc** [comment]
In fact, if I were doing this AnimatedDecalGeometry today, I would just keep original triangles around the placement point, without cutting / interpolating, and just replace the UVs - unlike the way DecalGeometry does it.

**Comment by Mugen87** [comment]
If you are so confident about this topic, feel free to make a PR. I would love to see how much code you need to get this work for skinned and morphed meshes^^.

**Comment by makc** [comment]
I did not think about morph targets, I will admit that.

**Comment by mrdoob** [comment]
Well, we could just do skinned for now.

**Comment by makc** [comment]
@mrdoob sorry I do not feel like PR-ing, as it requires a lot of dedication to keep @Mugen87 happy :)

kek, ok actually I just do not feel like submitting half-assed code. BUT, in case _somebody else_ would like to put in some work, [here is the jsfiddle](https://jsfiddle.net/qucL8pos/) to get you started.

**Comment by makc** [comment]
[version](https://jsfiddle.net/9zf1t26L/) with larger decal and slower animation, so you could see something larger than 10 pixels

**Comment by Mugen87** [comment]
> as it requires a lot of dedication to keep @Mugen87 happy :)

It's not me, it's actually the user base.
